File: world/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int get_length(long number);
int get_digit(long number, int position);
int get_sum(long number, int length);
int is_visa(long number, int length);
int is_master(long number, int length);
int is_ae(long number, int length);

int main (void){
    long number = get_long("Number: ");
    int length = get_length(number);
    int sum = get_sum(number, length);

    if(sum % 10 == 0){
        if(is_visa(number, length) == 0){
            printf("VISA\n");
        }else if(is_master(number, length) == 0){
            printf("MASTERCARD\n");
        }else if(is_ae(number, length) == 0){
            printf("AMEX\n");
        }else{
            printf("INVALID\n");
        }
    }else{
        printf("INVALID\n");
    }
}

int is_ae(long number, int length){
    if (length == 15){
        if(get_digit(number, length) == 3){
            if(get_digit(number, length - 1) == 4 || get_digit(number, length - 1) == 7){
                return 0;
            }else{
                return 1;
            }
        }else{
            return 1;
        }
    }else{
        return 1;
    }
}

int is_visa(long number, int length){
    if (length == 13 || length == 16){
        if(get_digit(number, length) == 4){
            return 0;
        }else{
            return 1;
        }
    }else{
        return 1;
    }
}

int is_master(long number, int length){
    if (length == 16){
        if(get_digit(number, length) == 5){
            if(get_digit(number, length - 1) == 1 || get_digit(number, length - 1) == 2 || get_digit(number, length - 1) == 3 || get_digit(number, length - 1) == 4 || get_digit(number, length - 1) == 5){
                return 0;
            }else{
                return 1;
            }
        }else{
            return 1;
        }
    }else{
        return 1;
    }
}

int get_length(long number){
    int length = 0;
    while(number > 0){
        number /= 10;
        length ++;
    }
    return length;
}

int get_digit(long number, int position){
    while (position > 1){
        number /= 10;
        position --;
    }

    int digit = number % 10;
    return digit;
}


int get_sum(long number, int length){
    int digit = 0;
    for(int pos_even = 2; pos_even <= length; pos_even += 2){
        int d = get_digit(number, pos_even) * 2;
        int l = get_length(d);
        if(l > 1){
            for(int pos = 1; pos <= l; pos ++){
                digit += get_digit(d, pos);
            }
        }else{
            digit += d;
        }
    }
    for(int pos_odd = 1; pos_odd <= length; pos_odd += 2){
        digit += get_digit(number, pos_odd);
    }
    return digit;
}

File: sentimental-credit/credit.py
from cs50 import get_int

number = get_int("Number: ")
length = len(str(number))

def isVisa(number, length):
    if (length == 13 or length == 16) and str(number)[0] == "4":
        return True
    return False

def isAE(number, length):
    code = int(str(number)[0] + str(number)[1])
    codes = [34, 37]
    if length == 15 and code in codes:
        return True
    return False

def isMaster(number, length):
    code = int(str(number)[0] + str(number)[1])
    codes = range(51, 56)
    if length == 16 and code in codes:
        return True
    return False

def checkSum(number, length):
    sum = 0
    for i in range(length - 2, -1, -2):
        sum = sum + int(str(number)[i]) * 2 if len(str(int(str(number)[i]) * 2)) == 1 else sum + int(str(int(str(number)[i]) * 2)[0]) + int(str(int(str(number)[i]) * 2)[1])
    for i in range(length - 1, -1, -2):
        sum += int(str(number)[i])
    if sum % 10 == 0:
        return True
    else:
        return False

type = "VISA" if isVisa(number, length) else "MASTERCARD" if isMaster(number, length) else "AMEX" if isAE(number, length) else "INVALID"
if type != "INVALID":
    type = "INVALID" if not checkSum(number, length) else type

print(type)
